crates/autopilot/src/decoded_settlement.rs

```diff
@@ -254,6 +254,68 @@ impl DecodedSettlement {
             }
         })
     }
+
+    fn fee(
+        &self,
+        external_prices: &ExternalPrices,
+        order: &OrderExecution,
+        trade: &DecodedTrade,
+    ) -> Option<U256> {
+        let solver_fee = match &order.executed_solver_fee {
+            Some(solver_fee) => u256_to_big_rational(solver_fee),
+            None => {
+                // this should be a partial limit order
+                if !trade.flags.partially_fillable() {
+                    tracing::warn!("missing solver fee for non partial fee order");
+                    return None;
+                }
+
+                // get uniform prices
+                let sell_index = self
+                    .tokens
+                    .iter()
+                    .position(|token| token == &order.sell_token)?;
+                let buy_index = self
+                    .tokens
+                    .iter()
+                    .position(|token| token == &order.buy_token)?;
+                let uniform_sell_price = self.clearing_prices.get(sell_index).cloned()?;
+                let uniform_buy_price = self.clearing_prices.get(buy_index).cloned()?;
+
+                // get executed(adjusted) prices
+                let sell_index = trade.sell_token_index.as_u64() as usize;
+                let buy_index = trade.buy_token_index.as_u64() as usize;
+                let adjusted_sell_price = self.clearing_prices.get(sell_index)?;
+                let adjusted_buy_price = self.clearing_prices.get(buy_index)?;
+
+                // the logic is opposite to the code in function `custom_price_for_limit_order`
+                let fee = match trade.flags.order_kind() {
+                    OrderKind::Buy => {
+                        let executed_sell_amount = trade
+                            .executed_amount
+                            .checked_mul(uniform_buy_price)?
+                            .checked_div(uniform_sell_price)?;
+                        adjusted_buy_price.checked_sub(executed_sell_amount)?
+                    }
+                    OrderKind::Sell => {
+                        let sell_amount = adjusted_sell_price
+                            .checked_mul(uniform_buy_price)?
+                            .checked_div(uniform_sell_price)?;
```

Comment:
It seems odd to me that we can compute the "real" sell amount only with prices.
I mean we currently encode the adjusted prices simply by storing the adjusted token amounts so in that sense `adjusted_sell_price` is equal to `adjusted_buy_amount` but this is not the only way one could encode the prices so people not using the reference driver could do it differently.
In my mind what needs to be done here is something like this:
```rust
let received_buy_amount = trade.executed_amount * adjusted_sell_price / adjusted_buy_price;
let sell_amount_needed_with_ucp = received_buy_amount * uniform_buy_price / uniform_sell_price;
let fee = trade.executed_amount - sell_amount_needed_with_ucp;
```
These price conversions are tricky so I might have gotten it wrong. Maybe @harisang can check these conversion?

Reply:
I think you got it right. And for partial limit orders we need to rely on `executed_amount * prices` and not on full amount. Committed the changes. UT passes for `sell` case.